### PR TITLE
SISRP-16933, translate LDAP-group: students-ug-grace to role: recentStudent

### PR DESCRIPTION
**Pull request by johncrossman**
https://jira.berkeley.edu/browse/SISRP-16933

My added comments in `Berkeley::UserRoles` are a good summary for this PR. I encourage both @raydavis and @paulkerschen to review.

If Jeremy gets back to me with more LDAP groups to consider then I'll submit another PR.

### CODE REVIEW

File: app/models/berkeley/user_roles.rb [hunk 1]
```diff
@@ -45,29 +45,26 @@ def roles_from_ldap_affiliations(ldap_record)
...
+      roles = find_matching_roles(ldap_groups, {
+        'cn=edu:berkeley:official:students:all,ou=campus groups,dc=berkeley,dc=edu' => [:student],
+        'cn=edu:berkeley:official:students:graduate,ou=campus groups,dc=berkeley,dc=edu' => [:graduate],
+        'cn=edu:berkeley:official:students:undergrad,ou=campus groups,dc=berkeley,dc=edu' => [:undergrad]
+      })
+      # Look for exStudent and recentStudent roles if person is not currently a student. As an example, a graduate
+      # student - identified in the block above - might have recently been an undergrad. In that scenario we do not
+      # want to confuse matters by tagging the user as both 'graduate' and 'recentStudent'.
+      unless [:student, :undergrad, :graduate].any? { |s| roles[s] }
+        # A recentStudent is an exStudent therefore we add both roles when user is in LDAP group(s) below.
```
Comment (johncrossman) [inline app/models/berkeley/user_roles.rb]: @raydavis What do you think of this reasoning?  "A recentStudent is an exStudent therefore we add both roles..."

Comment (raydavis) [inline app/models/berkeley/user_roles.rb]: @johncrossman: Nice improvements on the code in general, but here I'd instead suggest not even checking for recentStudent _unless_ exStudent has already been set by the affiliations check. In other words, drop the code's pretense that the CalGroups membership check can be an autonomous equal partner to the affiliations check, make it explicitly a supplemant by changing the method to "merge_roles_from_ldap_groups(ldap_groups, roles)", and replace the "unless" clause by an "if roles.include[:exStudent]" check.

Basically, I don't trust the CalGroups list (or our understanding of it) enough to give it the deciding vote on "exStudent" status. Even working from the very limited evidence gathered at CLC-6212, I see a case that this PR's logic would miss. This is a current but not-fully-registered-student:

```
   :berkeleyeduaffiliations=>
    ["EMPLOYEE-STATUS-EXPIRED",
     "STUDENT-STATUS-EXPIRED",
     "AFFILIATE-TYPE-ADVCON-ATTENDEE",
     "STUDENT-TYPE-NOT REGISTERED"],
   :berkeleyeduismemberof=>
    ["cn=edu:berkeley:app:calmessages:affiliates,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:app:calnet:compare-all-to-default,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:official:all,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:app:auth-cas:webapp-default:webapp-default-base-allow-not-allow,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:app:auth-cas:webapp-default:webapp-default-access,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:official:current-campus-community-plus-grace,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:app:auth-cas:webapp-default:webapp-default-base-affiliations,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:app:auth-cas:webapp-default:webapp-default-base-plus-allow,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:app:auth-cas:auth-default,ou=campus groups,dc=berkeley,dc=edu",
     "cn=edu:berkeley:official:students:students-ug-grace,ou=campus groups,dc=berkeley,dc=edu"
	 ], ...
```

Comment (johncrossman) [inline app/models/berkeley/user_roles.rb]: A lotta sense that makes. Thank you, @raydavis. An amendment is in.

**Comment by paulkerschen** [comment]
Thanks, @johncrossman, for yet more tricky multiple-source-of-record work. This looks right.

**Pull request merged**